### ISSUE TITLE
Is it possible to implement newly batch() from react-redux as an effect combinator

### ISSUE THREAD

**Issue by LoicGoyet** [redux-saga/redux-saga#1886]
### Description
Hello everyone,
React-redux introduced in 7.0.0 [`batch()`](https://react-redux.js.org/api/batch#batch) which allow :

> "multiple actions dispatched outside of React only result in a single render update."

This is pretty awesome as it can avoid unnecessary re-renders and allow a better control of our UI.

But I was wondering if it was possible into redux-saga to implement an effect combinator that would take as parameter an array of `put` effect and would result to dispatch them one by one into a `batch` call ?

It could looks like something like that :

```js
yield allBatch([
  put(action1()),
  put(action2()),
])
```

To my knowledge, this kind of pattern doesn't work directly with redux-saga, but I think it could be really awesome add into the API.

Thanks

**Comment by Andarist** [comment]
Let me think about how we could support this in a proper manner, but currently it should be possible to achieve this with current APIs. 

I've created a codesandbox demonstrating this - https://codesandbox.io/s/redux-saga-example-qq77q . I can't confirm the batching behavior though even though all dispatched actions reach reducer within batch call. Maybe @markerikson could give us a hand and take a look at this sandbox?

**Comment by markerikson** [comment]
You'd probably want to use the React DevTools (or the `<Profiler>` API) to count the number of renders/commits.

**Comment by LoicGoyet** [comment]
Thanks for the answers @Andarist @markerikson. Seems to work fine (I havent checked yet with the profiler how many commits react made) !

From my understanding, the solution you gave @Andarist plugs the `batch` for every `all`yield, and I don't really know what to think about it.

Is there some edge cases where we want to parallel non-blocking effect without batching them ? I don't really know if in some edge cases we could want to `yield all()`without the batching.

What are your thoughts about it ?

**Comment by Andarist** [comment]
This change definitely might change your components' lifecycle hooks logic - but if it does I would say that the code which has changed was error-prone anyway, so this actually should lead you to less bugs overall. 

I can't see the strong need to have unbatched all, keep in mind that only sync dispatches will get batched treatment so it really shouldnt matter if your component receives multiple change notifications within the same stack frame or if it receives only one - after all actions get processed by your reducers etc.

**Comment by benjaminbours** [comment]
I agree with @LoicGoyet , it would be an interesting feature and It would save some renders.

**Comment by Andarist** [comment]
Ok, my previous demo seems to work. I've tweaked it a little bit (added a thunk example):
https://codesandbox.io/s/redux-saga-example-b8lh4 . We could publish this middleware, so anyone interested could plug it in. This cannot be the default - because saga alone is not aware of react-redux (or even redux as a library) and its runtime doesn't depend on any of those (we depend on `redux` only because of its `compose` utility).

The odd thing for me is that `mapStateToProps` are still being called multiple times in `batch`. This can potentially be a lot of wasted work. Couldn't we workaround this, @markerikson ?

**Comment by markerikson** [comment]
No, this is expected behavior.

`batch()` works at the React level.  Any React updates that are queued inside the `batch()` call / in the same event loop tick will result in a single render pass, vs multiple individual render passes.

However, each individual dispatch still results in all of the store subscribers being notified, and thus the `mapState` functions run.

Now, there _is_ some trickiness in there because of how v7 is implemented.  Specifically, what you'd see is that any "top-level" connected components have their `mapState` functions run immediately on dispatch, while any _nested_ connected components would likely have their `mapState` functions run only when the queued render pass happens.  This is because of the custom tiered subscription handling we implement to enforce top-down updates with `connect`.

**Comment by Andarist** [comment]
Yeah sure, that all is understandable. I wonder though if your exported batch couldn't be integrated more into react-redux to only run mapState in the end. It would obviously have to hook into you subscription mechanism, but it would potentially be doable and could save quite some work from being just wasted as with current behavior what matters for the final (only) render is only the latest mapState's result.

**Comment by markerikson** [comment]
That's actually where stuff like https://github.com/tappleby/redux-batched-subscribe comes into play, but that requires modifying the store setup with an enhancer.  That's outside the scope of React-Redux.

**Comment by Andarist** [comment]
This is actually subtly different from what I had in mind. The enhancer technique affects all subscribers - the react-redux batch could only affect its own subscribers.

**Comment by chaitu9916** [comment]
I don't see any difference in number of commits when i add the above mentioned middleware. Why would that be?

**Comment by markerikson** [comment]
For reference, I wrote a blog post discussing the differences in various batching approaches:

https://blog.isquaredsoftware.com/2020/01/blogged-answers-redux-batching-techniques/

**Comment by chaitu9916** [comment]
Will check it and thank you for all your videos and blog posts, they are super helpful.

**Comment by Andarist** [comment]
Also note that React 18 will actually come with automatic batching so it will become less of an issue for libraries like Redux Saga
